File: problem_10/problem_10.py
import math
import concurrent.futures
from functools import partial


def is_prime(n: int) -> bool:
    if n == 2:
        return True
    if n < 2 or n % 2 == 0:
        return False
    for i in range(3, int(math.sqrt(n)) + 1, 2):
        if n % i == 0:
            return False
    return True


def calculate(_from: int, _to: int) -> int:
    total = 0
    while _from <= _to:
        if is_prime(_from):
            total += _from
        _from += 1
    return total


MAX_VALUE = 2_000_000
THREADS = 50

size = MAX_VALUE // THREADS
childrens = []


def main():
    actual = 1
    total = 0
    with concurrent.futures.ProcessPoolExecutor() as executor:
        for _ in range(THREADS):
            print(f"Calculating from {actual} to {actual + size - 1}")
            childrens.append(
                executor.submit(
                    partial(calculate, _to=((actual + size) - 1)),
                    actual,
                )
            )
            actual += size

        for future in concurrent.futures.as_completed(childrens):
            total += future.result()

    print(total)


if __name__ == "__main__":
    main()

File: problem_10/problem_10.rs
use std::thread;

const MAX_VALUE: i64 = 2_000_000;
const THREADS: i64 = 50;

fn is_prime(n: i64) -> bool {

    if n == 2 {
        return true;
    }

    if n < 2 || n % 2 == 0 {
        return false;
    }

    for i in (3..=((n as f64).sqrt() as i64)).step_by(2) {
        if n % i == 0 {
            return false;
        }
    }
    true
}

fn calculate(mut from: i64, to: i64) -> i64 {
    let mut total: i64 = 0;
    while from <= to {
        if is_prime(from) {
            total += from;
        }
        from += 1;
    }
    total
}

fn main() {
    let mut childrens = vec![];
    let size = MAX_VALUE / THREADS;
    let mut actual = 1;
    for _ in 0..THREADS {
        childrens.push(thread::spawn(move || {
            println!("Calculating from {} to {}", actual, (actual + size) - 1);
            calculate(actual, (actual + size) - 1)
        }));
        actual += size;
    }
    let total: i64 = childrens
        .into_iter()
        .map(|child| child.join().unwrap())
        .sum();

    println!("Total: {}", total);
}
